### DISCUSSION TITLE
Ideas & Suggestions

### DISCUSSION THREAD

**Discussion by webfansplz** [vuejs/devtools#111] [Announcements]
Any ideas & suggestions always welcome 👇

**Comment by michealroberts** [comment]
@webfansplz Great little tool!

My one nice to have would be to have the devtools bar aligned either top, left, right or bottom. As currently it cuts over some important UI/UX.

Other than that, really cool!

**Reply by alexzhang1030** [reply]
You can move devtools by dragging the devtools widget

**Reply by michealroberts** [reply]
Did not know this, but still would be nice to have control over it’s starting position.

**Reply by Offlein** [reply]
This was incredibly confusing to me too. I [made a PR](https://github.com/vuejs/devtools-next/pull/528) to put it into the docs.

**Reply by Craigeniah** [reply]
I would prefer the ability to have the dev tools in their own window, or undocked essentially. 
Example would be how the browser dev tools does this. Screenshot from Chrome/Brave Browser DevTools. 
![image](https://github.com/user-attachments/assets/d7a177eb-4e98-4c68-9184-ce93c4e4c4ca)

**Comment by g-zhima** [comment]
The lines in the background and the lines on the page are a bit confusing
![image](https://github.com/vuejs/devtools-next/assets/36685782/ed615847-7e96-4ae8-a11e-5e3b42bcaf1d)

**Reply by alexzhang1030** [reply]
You can open a PR if you have any interest :)

**Comment by DaniilIsupov** [comment]
`Performance` section is missing

**Reply by webfansplz** [reply]
The reason it is not supported yet is that it has performance issues, we will support it when we find a solution.

**Reply by JonathanBout** [reply]
@webfansplz I think you should enable the performance analyzer to analyze the performance analyzer's performance 😆

**Comment by michalkuncio** [comment]
I would love to see the Tanstack Query/Vue Query tab integrated inside the new Vue DevTools. I think in the "old/current" DevTools it was automatically detected and supported.

**Reply by webfansplz** [reply]
Let's track it in https://github.com/vuejs/devtools-next/discussions/150#discussioncomment-8013470.

**Comment by ChrisExP** [comment]
Is there a way to avoid it activating a button/link when trying to access the component data of that element?

**Comment by glancel** [comment]
Is there a way to see Component Events?

**Reply by pedro-pmda** [reply]
I would like to see the events as in the Dev Tools extension. When working with third party libraries (e.g. Vuetify) it is really interesting to track them.

**Comment by dansurlaru96** [comment]
I suggest you for this beautiful DevTools Tool to add plugin and modules section like a Nuxt.JS DevTools. See idea from this image. 
Thank you for support
<img width="1284" alt="217670797-12c33a03-ca4f-490d-a18a-ab9008b89c15" src="https://github.com/vuejs/devtools-next/assets/151963327/a940e25e-087c-480e-aa37-48eaa77cfd4c">

**Comment by skuhnow** [comment]
Maybe make the Dark/Light setting ("Settings/Appearance") not "dark" by default. It took quite some time to understand where the "dark" class in my root <html> element comes from. My suggestion: "None (or OS?), Light, Dark" with "None" as default

**Reply by webfansplz** [reply]
Let's track it in https://github.com/vuejs/devtools-next/issues/233

**Comment by jmanchenos** [comment]
it would be nice to have a filter on Component Tab to filter props, computed, etc by a string

**Reply by webfansplz** [reply]
I'll support it.

**Comment by Dykam** [comment]
A way to (always or easily) open it in a new window.

Our scenario has us use the tool in a very constrained space (addin to an application). What works is using `window.open("/__devtools__/")`, but having to do this every time at the start isn't too ergonomic.

And the normal floating panel isn't very useful when using the separate devtool window, maybe automatically hiding it could be good. Though not really necessary, the setting works tool.

Stand alone with `devtools.connect()` would be an obvious alternative, but for some reason it frequently disconnects, which this method doesn't suffer from.

**Reply by WebMechanic** [reply]
You could put that command into a bookmarklet so it's just a click.
Create a new bookmark, add this as its URL: `javascript:void( window.open("/__devtools__/") )`

**Comment by SimonEast** [comment]
Sometimes I would like to interact with a component in the developer console, calling methods and updating properties. It would be great if **clicking on a component in the "Components" tab would store a reference to the component instance in a global variable like `$v`** or something like that. This is similar to how Chrome stores a DOM element reference in `$0` when clicking on an HTML element in the "Elements" pane.

(I guess it would be great to allow console commands directly within the Vue Devtools pane itself, especially useful on mobile devices without a console, but I can also see that getting quite complex if you wanted autocomplete etc., so the request above is a lot simpler to implement.)

**Reply by skirtles-code** [reply]
This was implemented in #427.

You can access the currently selected component via `$vm` in the console. You can also use `$vm0`, ..., `$vm9` to access the last 10 components.

**Comment by alexzhang1030** [comment]
A small `codemod` tool for easy migration to new features.

Although the syntax of Vue3 hasn’t undergone significant changes, there are still some functional improvements (such as new features) between minor versions. We can write a simple codemod tool to help users easily migrate their syntax.

For example, Vue3.4 introduced a new feature `short-name v-bind`, and stabilized `defineModel`. By using `devtools-codemod` they can seamlessly migrate to new syntaxes without manually updating any code.

At the same time, this tool will also evolve with the iterations of Vue, presenting the new features of the latest Vue version to users in a more intuitive way.

**Comment by thedamon** [comment]
Would be really nice if there was a way to enable it to run in 'prod'/'build' mode. This may be very challenging or a non-starter because of the dependency on vite dev server, but with the old/current vue devtools browser plugin [you can make it run on prod with some console commands](https://mokkapps.de/vue-tips/force-enable-vue-devtools-in-production-build) and this can be extremely valuable

**Comment by stevecyj** [comment]
Is it possible to display the relative path of the file when hovering on open in editor?

**Comment by ghufrannoor2** [comment]
Could we display the current value of the router history?

It would be useful for debugging when dealing with issues where modifying the history doesn't work as expected, such as when implementing back/forward behavior.

**Reply by alexzhang1030** [reply]
> Could we display the current value of the router history?
> 
> It would be useful for debugging when dealing with issues where modifying the history doesn't work as expected, such as when implementing back/forward behavior.

Probably we can implement this feature on `vue-router`

**Comment by mohamad68** [comment]
plz add open in editor button 
![image](https://github.com/vuejs/devtools-next/assets/16140647/87312428-0e34-4f2d-88ec-8bb26952d363)

**Reply by alexzhang1030** [reply]
We do have it on `vite-plugin` because this feature relies on `vite middleware`

![image](https://github.com/vuejs/devtools-next/assets/49969959/b84bbdb0-fba2-4904-a409-2cb973ee6c68)

**Reply by webfansplz** [reply]
We'll follow the implementation of the old devtools to support it.

**Reply by avxkim** [reply]
@webfansplz can we expect this feature in this month?

**Comment by erk-dnr** [comment]
It would be nice to be able to see a visualization of component re-renders just like the React Devtools does that

**Comment by rs3d** [comment]
We're using multiple vue apps on a "static" CMS page:
![image](https://github.com/user-attachments/assets/a0fdc6f0-91ce-45cc-a6f6-c62261c03938)
It would be great to have a combined and searchable overview for apps and components, so you can find a specific one more easily.
Especially island architectures like Astro Islands would benefit from this change (https://docs.astro.build/en/concepts/islands/)

Maybe the component view could have <App/Root>-entries, if the website uses multiple ones.

**Comment by JonathanBout** [comment]
It'd be nice for the (already very cool!) graph to follow more references than just JS, for example the `import` statement in less. Currently separate less files just 'float around' even though they are attached via an import statement.

**Comment by niksy** [comment]
Option to disable animation when [inspecting components](https://devtools-next.vuejs.org/getting-started/features#inspector) on the page with "Select component on the page" button.

It can be very distracting jumping between large number of components on the small area.

Maybe take into account [reduced motion preference](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) and also have a toggle to force this option (similar to dark/light mode toggle in settings).

I suppose this is related to [vite-plugin-vue-inspector](https://github.com/webfansplz/vite-plugin-vue-inspector) so it’s best to start there with setting which can be toggled via devtools?

**Comment by zhangenming** [comment]
Visual vue responsive dependency graph

**Comment by rbgmulmb** [comment]
I am using JS map in **Pinia** storage. Currently **serialization** of Map doesn't work properly, it returns an empty object. It would be nice to add support for **Map** and **Set**. 
For this, special JSON replacers for Map and Set can be added.

**Comment by goper-leo** [comment]
Is there a keyboard shortcut for the "select component in the page"? A shortcut would be very helpful. Thanks.

**Comment by dawnlyn** [comment]
Would like to add multi-language support, e.g. Chinese

**Reply by alexzhang1030** [reply]
PR Welcome!

**Reply by RSS1102** [reply]
> PR Welcome!

good! If I have a chance I will try i18n work.

**Comment by zcrumbo** [comment]
My team is already really missing the ability to filter state in vuex/pinia

**Reply by webfansplz** [reply]
we’ll support it.

**Comment by renarsb** [comment]
bringing back highlighting components in the view when you select them would be nice

**Reply by alexzhang1030** [reply]
We have this on component panel

**Reply by renarsb** [reply]
it was more when you select/hover the node in the root tree it would highlight it in the view (if i have many components or if its not my project it is hards to navigate and understand which component is where).

**Reply by alexzhang1030** [reply]
![image](https://github.com/user-attachments/assets/7f286ddb-a609-4864-a196-a7b51fad717d)

You can use this as a temporary solution, as this might cause performance impacts, we will consider adding an option to support it.

**Comment by Craigeniah** [comment]
Moving this to it's own thread from my original comment at the top as it's slightly different from what was being suggested in that thread.
https://github.com/vuejs/devtools/discussions/111#discussioncomment-10896014

![image](https://github.com/user-attachments/assets/2db5dfa6-fe8a-42f7-af79-018961957a1c)

**Reply by alexzhang1030** [reply]
Well, we have electron, vite plugin, separate window and browser extension, use anyway you would prefer.

**Reply by Craigeniah** [reply]
Apologies, I guess I'm referring to the vite plugin. Normally I would just use it in the browser extension, and I was using it as the vite plugin and it would not allow me to undock it, just move it around the page.

**Reply by alexzhang1030** [reply]
You can try the separate window, I think that will perfectly solve your problem.

**Reply by Craigeniah** [reply]
> You can try the separate window, I think that will perfectly solve your problem.

Thanks for the info, can you elaborate on that a bit more however, where do I find that option?
I feel like I'm misunderstanding something.

**Reply by alexzhang1030** [reply]
<img width="1117" alt="image" src="https://github.com/user-attachments/assets/28aeea12-2399-4890-9ae4-0435bfbdafc7">

Sorry for delay, you can try this(you need open your page first)

**Reply by Craigeniah** [reply]
Thanks @alexzhang1030 I'll try that, appreciate the assist here!

**Comment by puyt** [comment]
In the previous version, you could toggle highlights on DOM elements as they (re)rendered, which was quite useful for debugging. 

This feature allowed you to see if a certain reactivity was updating more than one element, such as a list of items that should only update a single item when something changes, rather than all the items. 

I can't seem to find this toggle in the new devtools. Is it still available, or is it something that's planned for future updates?

**Reply by alexzhang1030** [reply]
Planned, maybe soon!

**Reply by puyt** [reply]
@alexzhang1030 is this still planned?

**Reply by alexzhang1030** [reply]
Hey, I’ve been quite busy with work lately, so I don’t have much time to tackle this right now. 😅 PRs are welcome, or I’ll get to it when I have some free time. Thanks for your patience! 🙌

**Reply by alexzhang1030** [reply]
#827

**Reply by puyt** [reply]
No problem, I totally understand.  I just wanted to check if this is still somewhere on the roadmap, as this previously existing feature would have been useful for a current use case I have.

Thanks for the update! 🙂

**Comment by fmgrafikdesign** [comment]
Please consider removing the "Welcome to the new version..." screen soon. When using incognito mode, the screen gets shown everytime on new browser contexts, which adds up to a lot of unnecessary extra clicks needed to actually use the tool.

**Reply by RSS1102** [reply]
Maybe should add a "skip all beginner guides"

**Comment by Martijn91** [comment]
Could we re-enable travelling the component tree with the use of arrow buttons again? This was very usefull with navigating through some nested components :)

**Comment by BenEdelstain** [comment]
Is it possible to track all api calls by components (for example on page load) / sort by same call path etc ?

**Comment by varHarrie** [comment]
![image](https://github.com/user-attachments/assets/e0cda769-eea6-408d-af98-1fc4e3838b09)

https://github.com/varHarrie/vite-plugin-vue-inspector-plus

I make an extension for vue devtools, which can quickly view the component hierarchy relationship of the target element and quickly open the component code file through combo key and mouse click. This is very helpful for development and I hope it can be integrated into the official function.

**Reply by alexzhang1030** [reply]
> ![image](https://private-user-images.githubusercontent.com/14304201/398509895-e0cda769-eea6-408d-af98-1fc4e3838b09.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMDYxNTgsIm5iZiI6MTczNTEwNTg1OCwicGF0aCI6Ii8xNDMwNDIwMS8zOTg1MDk4OTUtZTBjZGE3NjktZWVhNi00MDhkLWFmOTgtMWZjNGUzODM4YjA5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDA1NTA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4NDgyNTM2NDcyNDY1ZTA0M2E0MTcxNzcyOTI2NmQ5Y2VjNzk3N2ZlYjAzNTdlMjA2ODQ1NjFmNWMwOTlmMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Prqqw0zC3c-8ciJRGZHkQ41Bnh9e2euDGsWK0PnLUH8)
> 
> [varHarrie/vite-plugin-vue-inspector-plus](https://github.com/varHarrie/vite-plugin-vue-inspector-plus)
> 
> I make an extension for vue devtools, which can quickly view the component hierarchy relationship of the target element and quickly open the component code file through combo key and mouse click. This is very helpful for development and I hope it can be integrated into the official function.

Hi, do you have any interest in making a PR?

**Reply by varHarrie** [reply]
I'll review it when I have some free time, maybe submitting it to vite-plugin-vue-inspector would be better.

**Comment by malditochecho** [comment]
In the components tab, I would like to be able to navigate through the components with up/down keyboard arrows, and expand/collapse the tree of components with the left/right arrow keys.

Maybe also every time that a route is visited, automatically expand the tree up to the route's component, so its already selected and ready to be explored on the right panel.

Thanks!

**Comment by PeterCerven** [comment]
Not sure if this is an inspector issues or Vue Devtools issue but is there a way to trigger inspector while dialog window is open?

**Reply by BugBust86** [reply]
开发者工具真好用

**Comment by kimhornung** [comment]
I always go directly to "Components" after clicking the Vue tab in Dev Tools. For me, it would save a click if there was a setting to specify what the initial screen should be (instead of always opening with the overview, even when the overview has been disabled in Settings).

**Reply by LTDakin** [reply]
Came here to comment the same thing, the info page is fine, but even when I toggle it off It still shows up when I open the dev tools! And then I have to click components. Would be a great UX improvement

**Reply by luckydonald** [reply]
Additionally, I always click the locate icon, thinking it would lead me to that component in the inspector,
<img width="203" height="69" alt="image" src="https://github.com/user-attachments/assets/cd848ef5-402e-470b-a57f-ef83c09a9e2f" />
 but instead it opens my editor with the file of that component.

So basically, I'd rather have the following button there:
<img width="402" height="107" alt="image" src="https://github.com/user-attachments/assets/a868071f-2f3e-4491-b0ac-f97d4ea1d147" />
That page could then have a open in editor button.

In either way for opening the editor I'd rather use a share icon <img height="20" alt="image" src="https://github.com/user-attachments/assets/848145d0-28fb-4cc3-9ec8-495aed2c6b44" />  or a editor symbol  📝, not a sniper cross.

**Comment by Naedri** [comment]
Would it be possible to provide an iife version to the [npm repository](https://www.npmjs.com/package/@vue/devtools-api) ?
Indeed, it may help to close the [vuejs/pinia#2914](https://github.com/vuejs/pinia/issues/2914) issue.

**Comment by eleztian** [comment]
Can you consider supporting MCP, so that the cursor can make development more smooth?

**Reply by slavco86** [reply]
Was about to suggest something similar. That coupled with Playwright MCP would be amazing for AI debugging

**Reply by NikhilVerma** [reply]
Yes its a pain because AI tools can't debug or inspect Vue component state

**Comment by nuno-am** [comment]
Would you consider adding a new panel to track a11y issues?

**Comment by mepeisen** [comment]
I am playing around with custom inspectors. Mainly to interact with [SSE](https://en.wikipedia.org/wiki/Server-sent_events) and simulating my events during development. Anyhow it would be great to manipulate the nodeActions of an inspector depending on the selected node. For example selecting node "A" would display actions "Simulate Order" and "Simulate Cancellation". Selecting node "B" would display action "Simulate filling shopping cart".

I already see that this would break up API as well as the communication layer because getInspectorState should be able to send new nodeActions if needed. What do you think about it?

**Comment by Airkro** [comment]
Please release v7 on Edge extensions store. Chrome store Blocked by GFW.

**Comment by spaansba** [comment]
Can you consider adding a panel for [Tanstack Form](https://github.com/tanstack/form)? Would love to contribute if its something more people would find interesting.

**Comment by Caffeine19** [comment]
Add support for custom keyboard shortcuts to avoid conflicts with other apps or extensions.

**Comment by drilon-hyseni** [comment]
Hi. I would love the search bar in Pinia to automatically expand the objects and highlight the searched text. It would be a good functionality. 
Currently it filters only the object, but if there are way too many values under the expand sections, you cannot tell it.  
<img width="756" height="335" alt="image" src="https://github.com/user-attachments/assets/3fdd94e2-6be8-4fb9-bd57-398ba6d3504e" />

**Comment by Caffeine19** [comment]
MCP Support. To let the agent get the component's state, such as data, props, and the Pinia store state.

**Reply by webfansplz** [reply]
It's in our plan

**Comment by jasonlewis** [comment]
Is there plans to integrate the Vue devtools with the Vite devtools or will they remain separate? Would be nice to have it all bundled under the one floating icon rather than having multiple icons floating around.

**Reply by webfansplz** [reply]
Coming soon 👀

**Reply by jasonlewis** [reply]
Phew, keen to have it all in one place! Thanks heaps!

**Reply by webfansplz** [reply]
Released in v9.0.0-beta.0, try it out~

**Reply by jasonlewis** [reply]
Awesome, looks to be working really well, thanks heaps!
